battle.go

```diff
@@ -1,115 +1,212 @@
-	if len(b.Players) < 2 {
-		log.Panicf("not enough players to start a battle")
+// Used to provide data to Agents such that the Agent can't maliciously modify the state of the battle.
+type BattleInfo interface {
+	GetPokemon(int) Pokemon
+	Self() int
+	Allies() []int
+	Opponents() []int
+}
+
+// Implements the BattleInfo interface. Passed to Agents so they can decide a turn.
+type battleContext struct {
+	Context       Battle
+	ActivePokemon []activePokemon
+	self          int
+}
+
+func (c *battleContext) getTeam() []int {
+	selfAgent := c.ActivePokemon[c.self].AgentIdx
+	var team []int
+	for _, t := range c.Context.teams {
+		if contains(t, selfAgent) {
+			team = t
+			break
+		}
 	}
-	if len(b.Players[PLAYER1].Party) < 1 || len(b.Players[PLAYER2].Party) < 1 {
-		log.Panicf("not enough Pokemon in player parties to start a battle")
+	return team
+}
+func (c *battleContext) GetPokemon(idx int) Pokemon {
+	ap := c.ActivePokemon[idx]
+	b := c.Context
+	p := (*b.Parties[b.agentParties[ap.AgentIdx]]).Pokemon[ap.PokemonIdx]
+	return *p
+}
+func (c *battleContext) SetSelf(idx int) {
+	c.self = idx
+}
+func (c *battleContext) Self() int {
+	return c.self
+}
+func (c *battleContext) Allies() []int {
+	team := c.getTeam()
+
+	allies := []int{}
+	for i, ap := range c.ActivePokemon {
+		if i != c.self && contains(team, ap.AgentIdx) {
+			allies = append(allies, i)
+		}
 	}
-	b.Simulate()
-}
-
-// Simulates a battle using an event loop to handle all I/O and event dispatching
-func (b *Battle) Simulate() {
-	b.Dispatch(BATTLE_START)
-	for b.State != BATTLE_END {
-		switch b.State {
-		case BATTLE_START:
-			// Initialize battle conditions (weather, abilities, etc)
-			b.Dispatch(BEFORE_PLAYER1_TURN)
-		case BEFORE_PLAYER1_TURN:
-			// action := b.Player1.HandleTurn()
-			// Fix: Handle action
-			// fmt.Println("player 1 action:", action)
-			b.Dispatch(AFTER_PLAYER1_TURN)
-		case AFTER_PLAYER1_TURN:
-			// Fix: Handle events that occur between turns
-			// This includes weather, status effects, abilities, swaps, etc
-			b.Dispatch(BEFORE_PLAYER2_TURN)
-		case BEFORE_PLAYER2_TURN:
-			// action := b.Player2.HandleTurn()
-			// Fix: Handle action
-			// fmt.Println("player 2 action:", action)
-			b.Dispatch(AFTER_PLAYER2_TURN)
-		case AFTER_PLAYER2_TURN:
-			// Fix: Handle events that occur after both players move
-			// Also check if battle is over, otherwise go back to player 1
-			// Temporarily end battle after 1 round
-			if true {
-				b.Dispatch(BATTLE_END)
-				break
-			}
-			b.Dispatch(BEFORE_PLAYER1_TURN)
+
+	return allies
+}
+func (c *battleContext) Opponents() []int {
+	team := c.getTeam()
+
+	opponents := []int{}
+	for i, ap := range c.ActivePokemon {
+		if !contains(team, ap.AgentIdx) {
+			opponents = append(opponents, i)
 		}
 	}
-	// Cleanup after battle is over / ask to start new battle
+
+	return opponents
+}
+
+// An abstration over all possible actions an `Agent` can make in one round. Each Pokemon gets one turn.
```

Comment:
Continuing on my perspective from earlier, Turn should be a part of the Agent rather than on the Pokemon.